### PR TITLE
Migrate to Java 21

### PR DESCRIPTION
**Pull request by ravinperera00**
## Purpose

Fixes:

## Checklist
- [ ] Linked to an issue
- [ ] Updated the changelog
- [ ] Added tests
- [ ] Updated the spec

### CODE REVIEW

File: ballerina/tests/pubsub_test.bal [hunk 1]
```diff
@@ -41,10 +41,10 @@ function testGracefulShutdown() returns error? {
     stream<string, error?> subscribe = check pubsub.subscribe("topic");
     check pubsub.gracefulShutdown();
     record {|string value;|}|error? msg = subscribe.next();
-    if msg !is () {
-        test:assertFail(string `Expected "()", received ${(typeof msg).toString()}`);
+    test:assertTrue(msg is error);
+    if msg is error {
+        test:assertEquals(msg.message(), "Events must not be consumed from a closed pipe");
```
Comment (warunalakshitha) [inline ballerina/tests/pubsub_test.bal:46]: Why we need to change the tests?

Comment (ravinperera00) [reply ballerina/tests/pubsub_test.bal:46]: As far as I can understand, this is due to changes that has happened recently in [module-pipe](https://github.com/xlibb/module-pipe) and other related dependencies. But these changes has not yet been updated in `module-pubsub`. @Nuvindu might be able to better explain this and suggest alterations if needed.

File: native/src/main/java/io/xlibb/pubsub/PubSub.java [hunk 2]
```diff
@@ -62,16 +62,19 @@ public static Object subscribe(Environment environment, BObject pubsub, BString
...
-                                                             TypeCreator.createUnionType(PredefinedTypes.TYPE_ERROR,
-                                                                                         PredefinedTypes.TYPE_NULL));
+        Object[] arguments = new Object[]{timeout, typeParam};
+        CompletableFuture<Object> future = new CompletableFuture<>();
         MethodCallback callback = new MethodCallback(future);
-        environment.getRuntime()
-                .invokeMethodAsyncConcurrently(pipe, CONSUME_STREAM_METHOD, null, null, callback, null, streamType,
-                                               arguments);
-        return null;
+        return environment.yieldAndRun(() -> {
```
Comment (warunalakshitha) [inline native/src/main/java/io/xlibb/pubsub/PubSub.java]: Should we use a new thread here?

Comment (ravinperera00) [reply native/src/main/java/io/xlibb/pubsub/PubSub.java]: Fixed with 632ea15b67da0265c250a0b79b2860ba0151ab94

File: ballerina/tests/pubsub_test.bal (hunk 1 above)
Comment (Nuvindu) [reply ballerina/tests/pubsub_test.bal:46]: We updated an error message in the pipe package, but the `pubsub` module hasn't yet been released with that pipe version, so this fix is needed.

**Review by warunalakshitha — APPROVED** [review]

**Review by Nuvindu — APPROVED** [review]

**Pull request merged**